fork-runner/src/main/java/com/shazam/fork/system/PermissionGrantingManager.java

```diff
@@ -1,56 +1,96 @@
+import com.android.ddmlib.IDevice;
+import com.android.ddmlib.NullOutputReceiver;
+import com.android.ddmlib.ShellCommandUnresponsiveException;
+import com.android.ddmlib.TimeoutException;
 import com.shazam.fork.Configuration;
 import com.shazam.fork.model.Permission;
 
-import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
 
+import javax.annotation.Nonnull;
+
+import static com.shazam.fork.injector.ConfigurationInjector.configuration;
 import static java.lang.String.format;
 
 public class PermissionGrantingManager {
 
     private static final NullOutputReceiver NO_OP_RECEIVER = new NullOutputReceiver();
 
-    private PermissionGrantingManager() {
+    private final Logger logger = LoggerFactory.getLogger(PermissionGrantingManager.class);
+    private final Configuration configuration;
+
+    private PermissionGrantingManager(@Nonnull Configuration configuration) {
+        this.configuration = configuration;
     }
 
-    public static PermissionGrantingManager permissionGrantingManager(){
-        return new PermissionGrantingManager();
+    public static PermissionGrantingManager permissionGrantingManager() {
+        return new PermissionGrantingManager(configuration());
     }
 
     public void revokePermissions(@Nonnull String applicationPackage,
-                                         @Nonnull IDevice device, @Nonnull List<String> permissionsToRevoke) {
-        for (String permissionToRevoke : permissionsToRevoke) {
-            try {
-                device.executeShellCommand(format("pm revoke %s %s",
-                        applicationPackage, permissionToRevoke), NO_OP_RECEIVER);
-            } catch (TimeoutException | AdbCommandRejectedException | ShellCommandUnresponsiveException | IOException e) {
-                throw new UnsupportedOperationException(format("Unable to revoke permission %s", permissionToRevoke), e);
+                                  @Nonnull IDevice device,
+                                  @Nonnull List<String> permissionsToRevoke) {
+        if (!permissionsToRevoke.isEmpty()) {
+            long start = System.currentTimeMillis();
+            for (String permissionToRevoke : permissionsToRevoke) {
+                try {
+                    device.executeShellCommand(format("pm revoke %s %s",
+                            applicationPackage, permissionToRevoke), NO_OP_RECEIVER);
+                } catch (TimeoutException | AdbCommandRejectedException | ShellCommandUnresponsiveException | IOException e) {
+                    throw new UnsupportedOperationException(format("Unable to revoke permission %s", permissionToRevoke), e);
+                }
+            }
+
+            logger.debug("Revoking permissions: {} (took {}ms)", permissionsToRevoke, (System.currentTimeMillis() - start));
+        }
+    }
+
+    public void grantPermissions(@Nonnull String applicationPackage,
```

Comment:
I think this could be private now, and moved at the bottom of this class

Reply:
👍